src/functions/bitcoin/bitcoin-functions.ts

```diff
@@ -58,6 +60,51 @@ export function removeDustOutputs(
   }
 }
 
+export function getDerivedUnspendablePublicKeyCommittedToUUID(
+  vaultUUID: string,
+  bitcoinNetwork: Network
+): Buffer {
+  return deriveUnhardenedPublicKey(
+    getUnspendableKeyCommittedToUUID(vaultUUID, bitcoinNetwork),
+    bitcoinNetwork
+  );
+}
+
+export async function getVaultFundingBitcoinAddress(
+  vault: RawVault,
+  feeRecipient: string,
+  extendedAttestorGroupPublicKey: string,
+  bitcoinNetwork: Network,
+  bitcoinBlockchainAPIURL: string
+): Promise<string | undefined> {
+  const fundingTransaction = await fetchBitcoinTransaction(
+    vault.fundingTxId,
+    bitcoinBlockchainAPIURL
+  );
+
+  const multisigAddress = createTaprootMultisigPayment(
+    getDerivedUnspendablePublicKeyCommittedToUUID(vault.uuid, bitcoinNetwork),
+    deriveUnhardenedPublicKey(extendedAttestorGroupPublicKey, bitcoinNetwork),
+    Buffer.from(vault.taprootPubKey, 'hex'),
+    bitcoinNetwork
+  ).address;
+
+  const feeRecipientAddress = getFeeRecipientAddress(feeRecipient, bitcoinNetwork);
+
+  const inputAddresses = fundingTransaction.vin.map(input => input.prevout.scriptpubkey_address);
+
+  if (inputAddresses.every(address => address === inputAddresses[0])) {
+    if (inputAddresses[0] !== multisigAddress) {
+      return inputAddresses[0];
+    }
+    return fundingTransaction.vout.find(
+      output => output.scriptpubkey_address !== feeRecipientAddress
+    )?.scriptpubkey_address;
+  }
+
+  return inputAddresses.find(address => address !== multisigAddress);
+}
```

Comment:
this is problematic. You're returning in multiple places. I find this function's logic hard to follow